src/disp_s1/solid_earth_tides.py

```diff
@@ -0,0 +1,247 @@
+from rasterio.coords import BoundingBox
+from rasterio.warp import transform_bounds
+from scipy.interpolate import RegularGridInterpolator
+
+from disp_s1._reference import ReferencePoint
+
+
+def load_raster_bounds_and_transform(
+    filename: Filename,
+) -> Tuple[BoundingBox, rasterio.crs.CRS, rasterio.Affine, Tuple[int, int]]:
+    """Load bounds, transform, CRS, and shape from a raster file."""
+    try:
+        with rasterio.open(filename) as src:
+            bounds = src.bounds
+            crs = src.crs
+            affine_transform = src.transform
+            height, width = src.shape
+        return bounds, crs, affine_transform, (height, width)
+    except Exception as e:
+        raise RuntimeError(f"Error loading raster file {filename}: {e}") from e
+
+
+def transform_bounds_to_epsg4326(
+    bounds: BoundingBox, crs: rasterio.crs.CRS
+) -> BoundingBox:
+    """Transform bounds to EPSG:4326 if CRS is not geographic."""
+    if not crs.is_geographic:
+        transformed_bounds = transform_bounds(crs, "EPSG:4326", *bounds)
+        return BoundingBox(*transformed_bounds)
+    return bounds
+
+
+def generate_atr(
+    bounds: BoundingBox, height: int, width: int, margin_degrees: float = 0.1
+) -> dict:
+    """Generate the ATR dictionary for pysolid grid."""
+    min_grid_size = 25
+    height_small = max(min_grid_size, height // 100)
+    width_small = max(min_grid_size, width // 100)
+    return {
+        "LENGTH": height_small,
+        "WIDTH": width_small,
+        "X_FIRST": bounds.left - margin_degrees,
+        "Y_FIRST": bounds.top + margin_degrees,
+        "X_STEP": (bounds.right - bounds.left + 2 * margin_degrees) / width_small,
+        "Y_STEP": -(bounds.top - bounds.bottom + 2 * margin_degrees) / height_small,
+    }
+
+
+def interpolate_set_components(
+    n: np.ndarray,
+    e: np.ndarray,
+    y_arr: np.ndarray,
+    x_arr: np.ndarray,
+    set_east: np.ndarray,
+    set_north: np.ndarray,
+    set_up: np.ndarray,
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Interpolate SET components to the unwrapped interferogram grid."""
+    points = np.stack((n, e), axis=-1)
+    return (
+        RegularGridInterpolator((y_arr, x_arr), set_east)(points),
+        RegularGridInterpolator((y_arr, x_arr), set_north)(points),
+        RegularGridInterpolator((y_arr, x_arr), set_up)(points),
+    )
+
+
+def run_solid_grid(az_time, lat, lon, res_lat, res_lon):
+    """Run solid_grid from pysolid."""
+    az_time = pd.to_datetime(az_time)
+    return solid_grid(
+        az_time.year,
+        az_time.month,
+        az_time.day,
+        az_time.hour,
+        az_time.minute,
+        az_time.second,
+        lat,
+        res_lat,
+        1,
+        lon,
+        res_lon,
+        1,
+    )
+
+
+def calculate_time_ranges(
+    reference_start_time: pd.Timestamp,
+    reference_stop_time: pd.Timestamp,
+    secondary_start_time: pd.Timestamp,
+    secondary_stop_time: pd.Timestamp,
+    atr: dict,
```

Comment:
```suggestion
    shape: tuple[int, int],
```

(right now you can't tell what you need to pass to this function without looking at the code- making this `shape` means you can look at the function declaration+types, or the docstring if we were to add that)